src/sample.rs

```diff
@@ -0,0 +1,436 @@
+    ///
+    /// [`PerfRecordSample`]: struct.PerfRecordSample.html
+    PERIOD = sys::bindings::perf_event_sample_format_PERF_SAMPLE_PERIOD,
+
+    /// Fill out the `stream_id` field of [`PerfRecordSample`] when sampling.
+    ///
+    /// [`PerfRecordSample`]: struct.PerfRecordSample.html
+    STREAM_ID = sys::bindings::perf_event_sample_format_PERF_SAMPLE_STREAM_ID,
+
+    /// Fill out the `raw_sample` field of [`PerfRecordSample`] when sampling.
+    ///
+    /// [`PerfRecordSample`]: struct.PerfRecordSample.html
+    RAW = sys::bindings::perf_event_sample_format_PERF_SAMPLE_RAW,
+
+    /// Fill out the `weight` field of [`PerfRecordSample`] when sampling.
+    ///
+    /// [`PerfRecordSample`]: struct.PerfRecordSample.html
+    WEIGHT = sys::bindings::perf_event_sample_format_PERF_SAMPLE_WEIGHT,
+}
+
+/// A set of PerfSampleType that is implemented using bit math.
+#[derive(Default)]
+pub struct PerfSampleTypeSet(pub sys::bindings::perf_event_sample_format);
+
+impl PerfSampleTypeSet {
+    /// Add the given PerfSampleType to the set
+    pub fn add(&mut self, sample_type: PerfSampleType) {
+        self.0 |= sample_type as sys::bindings::perf_event_sample_format;
+    }
+
+    /// Returns true if the set contains the given PerfSampleType
+    pub fn contains(&self, sample_type: PerfSampleType) -> bool {
+        self.0 & sample_type as sys::bindings::perf_event_sample_format != 0
+    }
+}
+
+/// This record indicates a throttle / unthrottle event.
+#[derive(Debug)]
+pub struct PerfRecordThrottle {
+    /// Timestamp of when the record was created
+    pub time: u64,
+    /// A unique ID. If the event is a member of an event group, the group leader ID is
+    /// returned.
+    pub id: u64,
+    /// A unique ID. Unlike the above field, this is always the actual ID and never the group
+    /// leader ID.
+    pub stream_id: u64,
+}
+
+impl PerfRecordThrottle {
+    fn decode(_attrs: &sys::bindings::perf_event_attr, data: &[u8]) -> Result<Self> {
+        let mut offset = 0;
+        Ok(Self {
+            time: data.read(&mut offset)?,
+            id: data.read(&mut offset)?,
+            stream_id: data.read(&mut offset)?,
+        })
+    }
+}
+
+#[test]
+fn decode_perf_record_throttle() {
+    let attrs = sys::bindings::perf_event_attr::default();
+    let record = PerfRecordThrottle::decode(
+        &attrs,
+        &[
+            1, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 0, 0, 0, 0,
+        ],
+    )
+    .unwrap();
+    assert_eq!(record.time, 1);
+    assert_eq!(record.id, 2);
+    assert_eq!(record.stream_id, 3);
+}
+
+/// This record is a sample of the current state of some process.
+///
+/// Corresponds to the anonymous struct under `PERF_RECORD_SAMPLE` in [`perf_event_open`][man] man
+/// page.
+///
+/// [man]: http://man7.org/linux/man-pages/man2/perf_event_open.2.html
+#[derive(Debug, Default, PartialEq, Eq)]
+pub struct PerfRecordSample {
```

Comment:
I'm currently working on [a `linux-perf-event-reader` crate](https://docs.rs/linux-perf-event-reader/) which has some complete code for parsing these records. It also supports "wrap-around" reading without copies, via [`RawData::Split`](https://docs.rs/linux-perf-event-reader/0.4.0/linux_perf_event_reader/enum.RawData.html).

Reply:
Oh, that's great!

One of the reasons I haven't merged this PR is that I am concerned about testability. I don't want to have long sections of detailed, technical code that I have no way to ensure coverage for. But this weekend I finally got around to adding an interface for mocking the underlying system calls, so we can write nice deterministic tests that cover all the edge cases. So hopefully we'll be able to make some progress here.